plugins/batterystat.koplugin/main.lua

```diff
@@ -0,0 +1,273 @@
+    self.awake = Usage:new(self.settings:readSetting("awake"))
+    self.sleeping = Usage:new(self.settings:readSetting("sleeping"))
+
+    -- Note: these fields are not the "real" timestamp and battery usage, but
+    -- the unaccumulated values.
+    self.charging_state = State:new(self.settings:readSetting("charging_state"))
+    self.awake_state = State:new(self.settings:readSetting("awake_state"))
+    -- Whether the device was suspending before current timestamp.
+    self.was_suspending = false
+    -- Whether the device was charging before current timestamp.
+    self.was_charging = PowerD:isCharging()
+end
+
+function BatteryStat:onFlushSettings()
+    self.settings:replace({
+        charging = self.charging,
+        decharging = self.decharging,
+        awake = self.awake,
+        sleeping = self.sleeping,
+        charging_state = self.charging_state,
+        awake_state = self.awake_state,
+    })
+    self.settings:flush()
+end
+
+function BatteryStat:accumulate()
+    if self.was_suspending then
+        -- Suspending to awake.
+        self.sleeping:append(self.awake_state)
+    else
+        -- Awake to suspending, time between self.awake_state and now should belong to awake.
+        self.awake:append(self.awake_state)
+    end
+    if self.was_charging then
+        -- Decharging to charging.
+        self.charging:append(self.charging_state)
+    else
+        self.decharging:append(self.charging_state)
+    end
+    self.awake_state = State:new()
+    self.charging_state = State:new()
+end
+
+function BatteryStat:dumpOrLog(content)
+    local file = io.open(self.dump_file, "a")
+    if file then
+        file:write(content .. "\n")
+        file:close()
+    else
+        logger.warn("Failed to dump output ", content, " into ", self.dump_file )
+    end
+end
+
+function BatteryStat:debugOutput(event)
+    if self.debugging then
+        self:dumpOrLog(event .. " @ " .. State:new():toString() ..
+                       ", awake_state " .. self.awake_state:toString() ..
+                       ", charging_state " .. self.charging_state:toString())
+    end
+end
+
+function BatteryStat:onSuspend()
+    self:debugOutput("onSuspend")
+    self.was_suspending = false
+    self:accumulate()
+end
+
+function BatteryStat:onResume()
+    self:debugOutput("onResume")
+    self.was_suspending = true
+    self:accumulate()
+end
+
+function BatteryStat:onCharging()
+    self:debugOutput("onCharging")
+    self.was_charging = false
+    self:dumpToText()
+    self.charging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onNotCharging()
+    self:debugOutput("onNotCharging")
+    self.was_charging = true
+    self:dumpToText()
+    self.decharging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onCallback()
```

Comment:
It's just because of the length of the function and the indentation. If you have a strong opinion, I can embed it. I remember you have mentioned the performance lost of a lambda expression in LuaJIT, but I cannot quite tell it impacts here.